pyialarm/pyialarm.py

```diff
@@ -48,73 +48,68 @@ def __init__(self, host, port=18034):
         self.sock = None
 
     async def ensure_connection_is_open(self) -> None:
+        self._ensure_socket_is_open()
+
         if self.sock is None or self.sock.fileno() == -1:
             self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
             self.sock.setblocking(False)
-        else:
-            return
 
-        self.seq = 0
-        try:
+            self.seq = 0
             loop = asyncio.get_running_loop()
-            await loop.sock_connect(self.sock, (self.host, self.port))
-        except (TimeoutError, OSError, ConnectionRefusedError) as err:
-            with contextlib.suppress(OSError):
-                if self.sock:
-                    self.sock.close()
-            error_message = "Connection to the alarm system failed"
-            raise ConnectionError(error_message) from err
-        except Exception as err:
-            with contextlib.suppress(OSError):
-                if self.sock:
-                    self.sock.close()
-            error_message = "An unexpected error occurred"
-            raise RuntimeError(error_message) from err
+            try:
+                await loop.sock_connect(self.sock, (self.host, self.port))
+            except (TimeoutError, OSError, ConnectionRefusedError) as err:
+                self._close_connection()
+                raise IAlarmConnectionError from err
+            except Exception:
+                self._close_connection()
+                raise
+        else:
+            log.debug("Socket is already connected.")
+
+    def _ensure_socket_is_open(self) -> None:
+        if self.sock is None or self.sock.fileno() == -1:
+            self._close_connection()
 
     def _close_connection(self) -> None:
-        if self.sock and self.sock.fileno() != 1:
+        if self.sock and self.sock.fileno() != -1:
             self.sock.close()
```

Comment:
The implementation of `_ensure_socket_is_open()` seems off? I don't think it's expected to call `_close_connection()` when we already ensured the `self.sock` is None or without an open fileno.

If I'm not mistaken, probably you wanted to do something like the following?
```python
def _is_socket_open(self) -> bool:
    return self.sock and self.sock.fileno() != -1

[...]

async def ensure_connection_is_open(self) -> None:
    if not _is_socket_open():
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        [...]

async def _receive(self):
    if not _is_socket_open():
        raise_connection_error("Socket is not open")
```